### PR TITLE
Fix #2034, Perform autoswap on last deployment when there are multiple webhook triggers

### LINKED ISSUE #2034: Autoswap logic can cause a build to be killed halfway through
Here is the scenario that causes this:

- Deployment request comes in and takes a while
- While it's processing a second deployment request comes in. Here, we just update the marker file to mark there there is more work to do
- The first deployment completes, and triggers a Swap
- But now, the background thread finds that there is more work to do ([here](https://github.com/projectkudu/kudu/blob/44570100217a560563ce577cebc8be8fba470a96/Kudu.Services/ServiceHookHandlers/FetchHandler.cs#L297-L298)), and loops back to start another deployment.
- While that happens, the swap takes effect and kills the second deployment

### PR DESCRIPTION
**Pull request by shrimpy**
Only last deployment would trigger autoswap, below deployment log are from a sequence of consecutive webhook triggeded deployment, and it end up two deployments, and  only the last deployment will trigger autoswap.

````
[
{
log_time: "2016-06-09T01:40:46.7414829Z",
id: "152386e8-3918-4082-85db-66c3ceeb9949",
message: "Updating submodules.",
type: 0,
details_url: null
},
{
log_time: "2016-06-09T01:40:46.8884516Z",
id: "98b8215c-56b6-4a8e-9a06-f00b55bceceb",
message: "Preparing deployment for commit id '5e5cc1dfc4'.",
type: 0,
details_url: null
},
{
log_time: "2016-06-09T01:40:47.170165Z",
id: "4000547a-cc5b-4870-b3e0-d534dbb84f55",
message: "Generating deployment script.",
type: 0,
details_url: "https://kudutry.scm.azurewebsites.net/api/deployments/5e5cc1dfc45e7779982907d6d3bcbf1f1aeb47ec/log/4000547a-cc5b-4870-b3e0-d534dbb84f55"
},
{
log_time: "2016-06-09T01:40:47.2651707Z",
id: "24335ab5-7768-44ec-b2e3-31db3ef57f09",
message: "Running deployment command...",
type: 0,
details_url: "https://kudutry.scm.azurewebsites.net/api/deployments/5e5cc1dfc45e7779982907d6d3bcbf1f1aeb47ec/log/24335ab5-7768-44ec-b2e3-31db3ef57f09"
},
{
log_time: "2016-06-09T01:42:02.8921909Z",
id: "62bd4945-3cde-4295-8468-a131c83b4aba",
message: "Running post deployment command(s)...",
type: 0,
details_url: null
},
{
log_time: "2016-06-09T01:42:03.0176962Z",
id: "fde50c62-cc0c-4f0f-a12e-ef3e6dd508c5",
message: "Deployment successful.",
type: 0,
details_url: null
},
{
log_time: "2016-06-09T01:42:05.6152969Z",
id: "2d14e0ed-432f-4cab-b15a-985afcda8b89",
message: "Requesting auto swap to slot - 'production' operation id - 'AUTOSWAPf495cba5-eca4-4b0b-8d3d-47cd551541a3' deployment id - '5e5cc1dfc45e7779982907d6d3bcbf1f1aeb47ec'",
type: 0,
details_url: null
}
]
````

````
[
{
log_time: "2016-06-09T01:39:08.6699873Z",
id: "63b879a7-fc59-45c7-818a-0da209a8a952",
message: "Updating submodules.",
type: 0,
details_url: null
},
{
log_time: "2016-06-09T01:39:08.763638Z",
id: "783ceb5b-d73e-421a-bfb2-1bfebce42d9d",
message: "Preparing deployment for commit id 'a74a0fd8b1'.",
type: 0,
details_url: null
},
{
log_time: "2016-06-09T01:39:09.0356331Z",
id: "1bad9fb3-cb96-46e6-911b-382cefe43719",
message: "Generating deployment script.",
type: 0,
details_url: "https://kudutry.scm.azurewebsites.net/api/deployments/a74a0fd8b1d1fbebad942a788ca67febe37467a8/log/1bad9fb3-cb96-46e6-911b-382cefe43719"
},
{
log_time: "2016-06-09T01:39:10.0201723Z",
id: "bc309b72-afdb-4225-9019-29a5ab76b486",
message: "Running deployment command...",
type: 0,
details_url: "https://kudutry.scm.azurewebsites.net/api/deployments/a74a0fd8b1d1fbebad942a788ca67febe37467a8/log/bc309b72-afdb-4225-9019-29a5ab76b486"
},
{
log_time: "2016-06-09T01:40:42.7940499Z",
id: "b8b8b33a-b2a8-4904-b926-f08ea7ab34a9",
message: "Running post deployment command(s)...",
type: 0,
details_url: null
},
{
log_time: "2016-06-09T01:40:43.0284176Z",
id: "340aaeb4-51d9-408d-b055-d0d792561bf6",
message: "Deployment successful.",
type: 0,
details_url: null
}
]
````

### CODE REVIEW

File: Kudu.Core/Deployment/AutoSwapHandler.cs [hunk 1]
```diff
@@ -63,15 +60,7 @@ public bool IsAutoSwapOngoing()
                 tracer.Trace("Jwt token is null");
                 return;
             }
-
-            // active deployment is always a success deployment
-            string lastDeploymentId = _deploymentStatusManager.ActiveDeploymentId;
-            if (string.Equals(currentDeploymetId, lastDeploymentId, StringComparison.OrdinalIgnoreCase))
```
Comment (shrimpy) [inline Kudu.Core/Deployment/AutoSwapHandler.cs]: This check is no longer valid. For the case of subsequence webhook trigger, we perform autoswap after all deployment is done. 

So remove this check.

File: Kudu.Core/Deployment/DeploymentManager.cs [hunk 2]
```diff
@@ -617,7 +617,15 @@ private DeployResult GetResult(string id, string activeDeploymentId, bool isDepl
                     {
                         await builder.Build(context);
                         builder.PostBuild(context);
-                        await _autoSwapHandler.HandleAutoSwap(id, context);
+
+                        // There are two code path will trigger autoswap (if enabled), Local git and Webhook
+                        // For Local git we would like to trigger autoswap once deployment is done
+                        // For Webhook, there might be a sequence of trigger, we only want to perform autoswap when last deployment is done, and this can only handle outside of the deployment
+                        // Hence "swapAfterDeploy" flag to control whether to perform swap within a deployment
+                        if (swapAfterDeploy)
```
Comment (shrimpy) [inline Kudu.Core/Deployment/DeploymentManager.cs]: This "if" check is for local git, where deployment trigger by post-receive hook.  

I am thinking if we should move "_autoSwapHandler.HandleAutoSwap" outside of DeploymentManger.  Just as what I did in this PR, for the case of FetchHandler, perform autoswap after deployment process. Then we will not need this if statement and "swapAfterDeploy" property

File: Kudu.Console/Program.cs [hunk 3]
```diff
@@ -160,7 +160,7 @@ private static int Main(string[] args)
             {
                 try
                 {
-                    deploymentManager.DeployAsync(gitRepository, changeSet: null, deployer: deployer, clean: false)
+                    deploymentManager.DeployAsync(gitRepository, changeSet: null, deployer: deployer, clean: false, needFileUpdate: true, swapAfterDeploy: true)
```
Comment (davidebbo) [inline Kudu.Console/Program.cs]: No clear looking at the code why we always swap after deploy. What if auto-swap is not enabled? Or maybe the flag doesn't mean what it sounds like?

**Comment by davidebbo** [comment]
:shipit:

**Pull request merged**